src/inmanta/deploy/work.py

```diff
@@ -294,6 +307,7 @@ def reset(self) -> None:
     def deploy_with_context(
         self,
         resources: Set[ResourceIdStr],
+        priority: TaskPriority,
```

Comment:
I'm very much in favor of making this kw-only like the other args.

Reply:
If we do add this as kw-only, what do you suggest we use as default?

Reply:
It can be kw-only without having a default.

Reply:
But you do make me realize that it's the only non-trivial required field, so perhaps having it positional is acceptable after all.

Reply:
I still prefer kw-only, but I'm not as strongly opinionated on it as in my first comment.

Reply:
Why do you prefer kw-only?

Reply:
1. It's more explicit when the method is called, which I prefer for non-trivial arguments to a method. When you see the method name, you'll think "ok, we need to tell it what to deploy". It makes sense for this to be positional, and it makes sense for this to be the first argument because it's trivially expected to be there. But the same does not go for the priority.

Consider a `def add(x: int, y: int) -> int`, where it's obvious that it expects two arguments. However, when you add a context argument, it's not trivially expected there, e.g. `def add(x: int, y: int, *, print_result: bool = False)` where calling `add(1, 2, True)` would be unintuitive because based on the name you don't expect a boolean argument there. `add(1, 2, print_result=True)` is obvious to the reader, `add(1, 2, True) is vague at best, often confusing, and error-prone (see 2) at worst.

Now, because the priority is an enum rather than a simple int, it will always be clear what sort of context we're adding. So this argument only partially applies, which is why I don't feel so strongly about it.

2. In case of multiple non-trivial context arguments, there is no meaningful ordering, so you just pick an arbitrary order. For one, this can often lead to the confusion from 1, but it can also make refactors more difficult. Suppose you want to add an argument later. Either you have to put it at the end, or you have to make sure to update all callers. We have had (more than once) issues with exactly this: we refactor a method with multiple args of the same type (some with defaults), and we forgot to update some callers, leading to unexpected behavior.

Again, in this case, there's only the one argument, and it doesn't have a default, so argument 2 also doesn't fully apply.

Overall, I'm fine to leave it as is here, but the above has led me try to apply the reasoning as a general principle when I create an interface. I.e. think about "where does this argument make sense semantically?", "how will this read when called?", ... E.g. is it a trivial primary argument, or a context argument (as in the example with `add`)?

Sometimes there are strong motivations, sometimes it's more of a feeling that kw-only fits the context better. This situation here is more that subjective case, hence I express my preference but otherwise leave it up to you.

Reply:
Thanks for the thorough reply. It's also something that I subconsciously think but reading it this way made perfect sense. I always try to do it when I'm the user of the function I try to do:  `add(1, 2, print_result=True)` but as a developer I was only using kw-only when I had optional arguments.